### ISSUE TITLE
Unify argument validation

### ISSUE THREAD

**Issue by spawnia** [nuwave/lighthouse#1751]
**What problem does this feature proposal attempt to solve?**

The native spec-compliant validation performed by `webonyx/graphql-php` (`!`, scalars) and Lighthouse/Laravel (`@rules`, `@validator`) run in different phases and return different error shapes.

This has lead to number of issues:
- https://github.com/nuwave/lighthouse/issues/345
- https://github.com/nuwave/lighthouse/issues/262

**Which possible solutions should be considered?**

```
type Query {
  foo(
    nonNull: ID!
    customScalar: CustomScalar
    required: String @rules(apply: ["required"]
    email: String @rules(apply: ["email"]
  ): ID
}

query EverythingWrong {
  # Missing nonNull and required
  foo(customScalar: "wrong-format", email: "not-an-email")
}
```

An optimal solution would combine all kinds of validation in a single, unified step and return a consistent format for which inputs failed validation.

```json
{
  "errors": [
    {
      "message": "Validation failed for the field [foo]",
      "extensions": {
        "validation": {
          "nonNull": ["The nonNull field is required."],
          "customScalar": ["The customScalar field must be a valid CustomScalar."],
          "required": ["The required field is required."],
          "email": ["The email field must be a valid email."]
        }
      }
    }
  ]
}
```

**Comment by spawnia** [comment]
@wimski has written up a great summary and potential solution for this issue in https://github.com/nuwave/lighthouse/issues/1830

**Comment by wimski** [comment]
> wimski has written up a great summary and potential solution for this issue in 1830

In all fairness, it's not so much a solution as yet, more the desired result the solution should yield. I've had a quick look in how GraphQL validation is done and I think it's going to be pretty hard to somehow hook into/extend that. But I'll definitely give it another go when I have a bit more time.

**Comment by spawnia** [comment]
Perhaps we need to swap out some of the validation rules in `webonyx/graphql-php` with our own implementation: https://github.com/webonyx/graphql-php/blob/bb60eb49bd77dd04b78ff7e54d0098d3f075d261/src/Validator/DocumentValidator.php#L140-L167

**Comment by wimski** [comment]
Yes, the `ValuesOfCorrectType` to be exact. But it's hard to extend with all the `private` and `static`/`self` going on.

**Comment by spawnia** [comment]
You can create a PR in `webonyx/graphql-php` to make those into `protected/static`.

**Comment by wimski** [comment]
Removing all the static would make things already a lot easier for sure.
https://github.com/webonyx/graphql-php/issues/851

**Comment by chojnicki** [comment]
3 years passed - @spawnia @wimski did you guys tried to work on it, with any results? All I want to do is to use "required" validation so it will return nice error handled by frontend.

**Comment by spawnia** [comment]
Nice idea @chojnicki, I never thought of modifying the native validation errors to make them look like Laravel validation errors. There are many other possible validation errors that would need to be added in order for this to be complete (e.g. all the scalar validations).

**Comment by fer-ri** [comment]
> use Illuminate\Validation\ValidationException as LaravelValidationException;

Thanks a lot for the idea, Here's mine

```php
<?php

namespace App\Exceptions\GraphQL;

use GraphQL\Error\Error;
use Illuminate\Validation\ValidationException as LaravelValidationException;
use Nuwave\Lighthouse\Exceptions\ValidationException;
use Nuwave\Lighthouse\Execution\ErrorHandler;

final class NonNullFieldErrorHandler implements ErrorHandler
{
    public function __invoke(?Error $error, \Closure $next): ?array
    {
        if (preg_match('/Variable "(.+)" got invalid value null at "(.+)"; Expected non-nullable type "(.+)" not to be null./', $error->getMessage(), $matches)) {
            $fieldName = $matches[2];

            $error = new Error(
                $error->getMessage(),
                $error->getNodes(),
                $error->getSource(),
                $error->getPositions(),
                $error->getPath(),
                ValidationException::fromLaravel(LaravelValidationException::withMessages([
                    $fieldName => trans('validation.required', ['attribute' => $fieldName]),
                ])),
            );
        }

        return $next($error);
    }
}
```
